src/mod_ping.erl

```diff
@@ -152,43 +149,62 @@ iq_ping(Acc, _From, _To, #iq{sub_el = SubEl} = IQ, _) ->
 %% Hook callbacks
 %%====================================================================
 
-handle_remote_hook(HandlerState, mod_ping, Args, C2SState) ->
-    handle_remote_call(Args,
+-spec handle_remote_hook(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: term(),
+    Params :: #{tag := atom(), args := term(), c2s_state := ejabberd_c2s:state()},
+    Extra :: map().
+handle_remote_hook(HandlerState, #{tag := mod_ping, hook_args := Args, c2s_state := C2SState}, _) ->
+    {ok, handle_remote_call(Args,
                        ejabberd_c2s_state:jid(C2SState),
                        ejabberd_c2s_state:server(C2SState),
                        ejabberd_c2s_state:host_type(C2SState),
-                       HandlerState);
-handle_remote_hook(HandlerState, _, _, _) ->
-    HandlerState.
-
-user_online(Acc, _HostType, {_, Pid} = _SID, _Jid, _Info) ->
+                       HandlerState)};
+handle_remote_hook(HandlerState, _, _) ->
+    {ok, HandlerState}.
+
+-spec user_online(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: ok,
+    Params :: #{sid := 'undefined' | ejabberd_sm:sid()},
```

Comment:
Is `'undefined'` necessary here? It will fail with a function_clause if so 🤔 Also the apostrophes aren't needed, it could be just `undefined`.

Reply:
I took it from `mongoose_hooks:sm_remove_connection_hook/5` and `mongoose_hooks:sm_register_connection_hook/4`'s spec. You're right though, for this handler it makes no sense. I handler fails we're catching error and proceed with other handlers, so I guess there is no need to implement such function clause. I'll simply remove it from spec.